// ao-angular/esm5/lib/services/ao-auth.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/ao-auth.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { EventEmitter, Injectable } from '@angular/core';
import { NavController } from '@ionic/angular';
import { LocalStorage } from 'ngx-webstorage';
import { BehaviorSubject } from 'rxjs';
import { AOService } from './ao.service';
import * as i0 from "@angular/core";
import * as i1 from "@ionic/angular";
import * as i2 from "./ao.service";
// @dynamic
var AOAuthService = /** @class */ (function () {
    // -------------------------------------------------------------------------------------------------------------------
    function AOAuthService(_nav, _ao) {
        var _this = this;
        this._nav = _nav;
        this._ao = _ao;
        this._redirect_default = '/';
        this.redirect = (/**
         * @return {?}
         */
        function () { return _this._redirect; });
        this.user = (/**
         * @return {?}
         */
        function () { return _this._user; });
        // -------------------------------------------------------------------------------------------------------------------
        this.isUser = (/**
         * @return {?}
         */
        function () { return !!_this._user; });
        this.isGuest = (/**
         * @return {?}
         */
        function () { return !_this._user; });
        this.id = (/**
         * @return {?}
         */
        function () { return _this.isUser() ? _this.user().id : null; });
        this.name = (/**
         * @return {?}
         */
        function () { return _this.isUser() ? _this.user().name : null; });
        this.email = (/**
         * @return {?}
         */
        function () { return _this.isUser() ? _this.user().email : null; });
        this.image = (/**
         * @return {?}
         */
        function () { return _this.isUser() ? _this.user().image : null; });
        this.verified = (/**
         * @return {?}
         */
        function () { return _this.isUser() ? _this.user().verified : null; });
        this.permitions = (/**
         * @return {?}
         */
        function () { return _this.isUser() ? _this.user().permitions : null; });
        // -------------------------------------------------------------------------------------------------------------------
        this.onLogin = (/**
         * @return {?}
         */
        function () { return AOAuthService._onLogin; });
        this.onLogout = (/**
         * @return {?}
         */
        function () { return AOAuthService._onLogout; });
        this.onChange = (/**
         * @return {?}
         */
        function () { return AOAuthService._onChange; });
        AOAuthService._onChange = new BehaviorSubject(this.user());
    }
    // -------------------------------------------------------------------------------------------------------------------
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?=} provider
     * @return {?}
     */
    AOAuthService.prototype.provider = 
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?=} provider
     * @return {?}
     */
    function (provider) {
        if (provider === void 0) { provider = null; }
        if (provider == null) {
            return this._provider;
        }
        this._provider = provider;
        return this;
    };
    // -------------------------------------------------------------------------------------------------------------------
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} redirect
     * @return {THIS}
     */
    AOAuthService.prototype.setRedirect = 
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} redirect
     * @return {THIS}
     */
    function (redirect) {
        (/** @type {?} */ (this))._redirect = redirect;
        return (/** @type {?} */ (this));
    };
    // -------------------------------------------------------------------------------------------------------------------
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} user
     * @return {THIS}
     */
    AOAuthService.prototype.setUser = 
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} user
     * @return {THIS}
     */
    function (user) {
        (/** @type {?} */ (this))._user = user;
        (/** @type {?} */ (this)).onChange().next((/** @type {?} */ (this))._user);
        return (/** @type {?} */ (this));
    };
    // -------------------------------------------------------------------------------------------------------------------
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} data
     * @return {?}
     */
    AOAuthService.prototype.login = 
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        return this._ao.wait(this.provider().login(data)).then((/**
         * @param {?} user
         * @return {?}
         */
        function (user) {
            _this.setUser(user);
            _this.onLogin().emit();
            /** @type {?} */
            var redirect = _this._redirect && _this._redirect !== '/login'
                ? _this._redirect
                : _this._redirect_default;
            _this._redirect = null;
            return _this._nav.navigateRoot(redirect);
        }));
    };
    /**
     * @return {?}
     */
    AOAuthService.prototype.logout = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return this._ao.wait(this.provider().logout()).then((/**
         * @return {?}
         */
        function () {
            _this.setUser(null);
            _this.onLogout();
            return _this._nav.navigateRoot('/login');
        }));
    };
    /**
     * @param {?} data
     * @return {?}
     */
    AOAuthService.prototype.update = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        return this._ao.wait(this.provider().update(data));
    };
    /**
     * @param {?} data
     * @return {?}
     */
    AOAuthService.prototype.updatePassword = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        return this._ao.wait(this.provider().updatePassword(data));
    };
    /**
     * @param {?} data
     * @return {?}
     */
    AOAuthService.prototype.forgetPassword = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        return this._ao.wait(this.provider().forgetPassword(data));
    };
    /**
     * @param {?} data
     * @return {?}
     */
    AOAuthService.prototype.resetPassword = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        return this._ao.wait(this.provider().resetPassword(data));
    };
    AOAuthService._onLogin = new EventEmitter();
    AOAuthService._onLogout = new EventEmitter();
    AOAuthService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    AOAuthService.ctorParameters = function () { return [
        { type: NavController },
        { type: AOService }
    ]; };
    /** @nocollapse */ AOAuthService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function AOAuthService_Factory() { return new AOAuthService(i0.ɵɵinject(i1.NavController), i0.ɵɵinject(i2.AOService)); }, token: AOAuthService, providedIn: "root" });
    tslib_1.__decorate([
        LocalStorage('user'),
        tslib_1.__metadata("design:type", Object)
    ], AOAuthService.prototype, "_user", void 0);
    tslib_1.__decorate([
        LocalStorage('redirect'),
        tslib_1.__metadata("design:type", String)
    ], AOAuthService.prototype, "_redirect", void 0);
    return AOAuthService;
}());
export { AOAuthService };
if (false) {
    /** @type {?} */
    AOAuthService._onLogin;
    /** @type {?} */
    AOAuthService._onLogout;
    /** @type {?} */
    AOAuthService._onChange;
    /**
     * @type {?}
     * @private
     */
    AOAuthService.prototype._user;
    /**
     * @type {?}
     * @private
     */
    AOAuthService.prototype._redirect;
    /**
     * @type {?}
     * @private
     */
    AOAuthService.prototype._redirect_default;
    /**
     * @type {?}
     * @private
     */
    AOAuthService.prototype._provider;
    /** @type {?} */
    AOAuthService.prototype.redirect;
    /** @type {?} */
    AOAuthService.prototype.user;
    /** @type {?} */
    AOAuthService.prototype.isUser;
    /** @type {?} */
    AOAuthService.prototype.isGuest;
    /** @type {?} */
    AOAuthService.prototype.id;
    /** @type {?} */
    AOAuthService.prototype.name;
    /** @type {?} */
    AOAuthService.prototype.email;
    /** @type {?} */
    AOAuthService.prototype.image;
    /** @type {?} */
    AOAuthService.prototype.verified;
    /** @type {?} */
    AOAuthService.prototype.permitions;
    /** @type {?} */
    AOAuthService.prototype.onLogin;
    /** @type {?} */
    AOAuthService.prototype.onLogout;
    /** @type {?} */
    AOAuthService.prototype.onChange;
    /**
     * @type {?}
     * @private
     */
    AOAuthService.prototype._nav;
    /**
     * @type {?}
     * @private
     */
    AOAuthService.prototype._ao;
}
