// ao-angular/esm5/lib/services/ao-auth-provider.interface.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/ao-auth-provider.interface.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function AOAuthProviderInterface() { }
if (false) {
    /**
     * @param {?} data
     * @return {?}
     */
    AOAuthProviderInterface.prototype.login = function (data) { };
    /**
     * @return {?}
     */
    AOAuthProviderInterface.prototype.logout = function () { };
    /**
     * @param {?} data
     * @return {?}
     */
    AOAuthProviderInterface.prototype.update = function (data) { };
    /**
     * @param {?} data
     * @return {?}
     */
    AOAuthProviderInterface.prototype.forgetPassword = function (data) { };
    /**
     * @param {?} data
     * @return {?}
     */
    AOAuthProviderInterface.prototype.resetPassword = function (data) { };
    /**
     * @param {?} data
     * @return {?}
     */
    AOAuthProviderInterface.prototype.updatePassword = function (data) { };
}
