// ao-angular/esm5/lib/services/ao-firebase-connection.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/ao-firebase-connection.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { AngularFireAuth } from '@angular/fire/auth';
import { AngularFireDatabase } from '@angular/fire/database';
import { AngularFireStorage } from '@angular/fire/storage';
import { AngularFirestore } from '@angular/fire/firestore';
import { AngularFireMessaging } from '@angular/fire/messaging';
var AOFirebaseConnectionService = /** @class */ (function () {
    function AOFirebaseConnectionService(_name, _hash, _app, _zone) {
        this._name = _name;
        this._hash = _hash;
        this._app = _app;
        this._zone = _zone;
        this._configs = {};
    }
    /**
     * @return {?}
     */
    AOFirebaseConnectionService.prototype.app = /**
     * @return {?}
     */
    function () {
        return this._app;
    };
    /**
     * @return {?}
     */
    AOFirebaseConnectionService.prototype.name = /**
     * @return {?}
     */
    function () {
        return this._name;
    };
    /**
     * @return {?}
     */
    AOFirebaseConnectionService.prototype.hash = /**
     * @return {?}
     */
    function () {
        return this._hash;
    };
    // -------------------------------------------------------------------------------------------------------------------
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @return {?}
     */
    AOFirebaseConnectionService.prototype.auth = 
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @return {?}
     */
    function () {
        return this._auth
            ? this._auth
            : this._auth = new AngularFireAuth(this._configs, this._name, null, this._zone);
    };
    /**
     * @return {?}
     */
    AOFirebaseConnectionService.prototype.database = /**
     * @return {?}
     */
    function () {
        return this._database
            ? this._database
            : this._database = new AngularFireDatabase(this._configs, this._name, null, null, this._zone);
    };
    /**
     * @return {?}
     */
    AOFirebaseConnectionService.prototype.firestore = /**
     * @return {?}
     */
    function () {
        return this._firestore
            ? this._firestore
            : this._firestore = new AngularFirestore(this._configs, this._name, null, null, null, this._zone, null);
    };
    /**
     * @return {?}
     */
    AOFirebaseConnectionService.prototype.messaging = /**
     * @return {?}
     */
    function () {
        return this._messaging
            ? this._messaging
            : this._messaging = new AngularFireMessaging(this._configs, this._name, null, this._zone);
    };
    /**
     * @return {?}
     */
    AOFirebaseConnectionService.prototype.storage = /**
     * @return {?}
     */
    function () {
        return this._storage
            ? this._storage
            : this._storage = new AngularFireStorage(this._configs, this._name, null, null, this._zone);
    };
    // -------------------------------------------------------------------------------------------------------------------
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @return {?}
     */
    AOFirebaseConnectionService.prototype.logout = 
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @return {?}
     */
    function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var e_1;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this._auth) return [3 /*break*/, 4];
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this._auth.auth.signOut()];
                    case 2:
                        _a.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        e_1 = _a.sent();
                        console.warn('Oops! Failed to run signOut.');
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/, Promise.resolve(true)];
                }
            });
        });
    };
    /**
     * @return {?}
     */
    AOFirebaseConnectionService.prototype.destroy = /**
     * @return {?}
     */
    function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.logout()];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this._app.delete()];
                    case 2:
                        _a.sent();
                        return [2 /*return*/, Promise.resolve(true)];
                }
            });
        });
    };
    return AOFirebaseConnectionService;
}());
export { AOFirebaseConnectionService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AOFirebaseConnectionService.prototype._configs;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseConnectionService.prototype._auth;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseConnectionService.prototype._database;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseConnectionService.prototype._firestore;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseConnectionService.prototype._messaging;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseConnectionService.prototype._storage;
    /**
     * @type {?}
     * @protected
     */
    AOFirebaseConnectionService.prototype._name;
    /**
     * @type {?}
     * @protected
     */
    AOFirebaseConnectionService.prototype._hash;
    /**
     * @type {?}
     * @protected
     */
    AOFirebaseConnectionService.prototype._app;
    /**
     * @type {?}
     * @protected
     */
    AOFirebaseConnectionService.prototype._zone;
}
