// ao-angular/esm5/lib/services/ao-firebase-factory.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/ao-firebase-factory.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Inject, Injectable, NgZone } from '@angular/core';
import { FirebaseApp } from '@angular/fire';
import { _firebaseAppFactory } from '@angular/fire/firebase.app.module';
import { AOFirebaseConnectionService } from './ao-firebase-connection.service';
import { AOEnvService } from './ao-env.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/fire";
import * as i2 from "./ao-env.service";
var AOFirebaseFactoryService = /** @class */ (function () {
    function AOFirebaseFactoryService(_app, _zone, _environment) {
        this._app = _app;
        this._zone = _zone;
        this._environment = _environment;
        this.connections = {};
        this.factory(this._environment.firebase.name, this._environment.firebase.configs, this._app);
    }
    /**
     * @param {?=} name
     * @return {?}
     */
    AOFirebaseFactoryService.prototype.conn = /**
     * @param {?=} name
     * @return {?}
     */
    function (name) {
        if (this.connections[name]) {
            return this.connections[name];
        }
        throw new Error('FirebaseApp(' + name + ') not found.');
    };
    /**
     * @param {?} configs
     * @return {?}
     */
    AOFirebaseFactoryService.prototype.makeHash = /**
     * @param {?} configs
     * @return {?}
     */
    function (configs) {
        return JSON.stringify({
            projectId: configs.projectId,
            apiKey: configs.apiKey
        });
    };
    /**
     * @param {?} name
     * @param {?=} configs
     * @param {?=} app
     * @return {?}
     */
    AOFirebaseFactoryService.prototype.factory = /**
     * @param {?} name
     * @param {?=} configs
     * @param {?=} app
     * @return {?}
     */
    function (name, configs, app) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var hash;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        hash = this.makeHash(configs);
                        if (!this.connections[name]) return [3 /*break*/, 3];
                        if (!(this.connections[name].hash() === hash)) return [3 /*break*/, 1];
                        return [2 /*return*/, Promise.resolve(true)];
                    case 1: return [4 /*yield*/, this.connections[name].destroy()];
                    case 2:
                        _a.sent();
                        _a.label = 3;
                    case 3:
                        if (!app) {
                            app = _firebaseAppFactory((/** @type {?} */ (configs)), this._zone, name);
                        }
                        this.connections[name] = new AOFirebaseConnectionService(name, hash, app, this._zone);
                        return [2 /*return*/, Promise.resolve(true)];
                }
            });
        });
    };
    /**
     * @param {?=} only
     * @param {?=} except
     * @return {?}
     */
    AOFirebaseFactoryService.prototype.destroy = /**
     * @param {?=} only
     * @param {?=} except
     * @return {?}
     */
    function (only, except) {
        if (only === void 0) { only = []; }
        if (except === void 0) { except = ['master']; }
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var promises;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                if (only.length === 0) {
                    only = Object.keys(this.connections);
                }
                promises = [];
                Object.keys(this.connections).forEach((/**
                 * @param {?} name
                 * @return {?}
                 */
                function (name) {
                    if (only.indexOf(name) >= 0 && except.indexOf(name) === -1) {
                        promises.push(_this.conn(name).destroy());
                    }
                }));
                return [2 /*return*/, Promise.all(promises).then((/**
                     * @return {?}
                     */
                    function () { return true; }))];
            });
        });
    };
    AOFirebaseFactoryService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    AOFirebaseFactoryService.ctorParameters = function () { return [
        { type: FirebaseApp },
        { type: NgZone },
        { type: undefined, decorators: [{ type: Inject, args: [AOEnvService,] }] }
    ]; };
    /** @nocollapse */ AOFirebaseFactoryService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function AOFirebaseFactoryService_Factory() { return new AOFirebaseFactoryService(i0.ɵɵinject(i1.FirebaseApp), i0.ɵɵinject(i0.NgZone), i0.ɵɵinject(i2.AOEnvService)); }, token: AOFirebaseFactoryService, providedIn: "root" });
    return AOFirebaseFactoryService;
}());
export { AOFirebaseFactoryService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AOFirebaseFactoryService.prototype.connections;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseFactoryService.prototype._app;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseFactoryService.prototype._zone;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseFactoryService.prototype._environment;
}
