// ao-angular/esm5/lib/services/ao-menu.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/ao-menu.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { MenuController, Platform } from '@ionic/angular';
import * as i0 from "@angular/core";
import * as i1 from "@ionic/angular";
var AOMenuService = /** @class */ (function () {
    function AOMenuService(platform, controller) {
        this.platform = platform;
        this.controller = controller;
        this._stacked = ['lg'];
    }
    /**
     * @return {?}
     */
    AOMenuService.prototype.disable = /**
     * @return {?}
     */
    function () {
        return this.controller.enable(false, 'AO_APP_MENU');
    };
    /**
     * @return {?}
     */
    AOMenuService.prototype.enable = /**
     * @return {?}
     */
    function () {
        return this.controller.enable(true, 'AO_APP_MENU');
    };
    /**
     * @return {?}
     */
    AOMenuService.prototype.stackable = /**
     * @return {?}
     */
    function () {
        return this.platform.width() > 991;
    };
    /**
     * @return {?}
     */
    AOMenuService.prototype.stacked = /**
     * @return {?}
     */
    function () {
        return this._stacked;
    };
    /**
     * @return {?}
     */
    AOMenuService.prototype.stackToggle = /**
     * @return {?}
     */
    function () {
        this._stacked = this._stacked == false ? ['lg'] : false;
    };
    AOMenuService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    AOMenuService.ctorParameters = function () { return [
        { type: Platform },
        { type: MenuController }
    ]; };
    /** @nocollapse */ AOMenuService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function AOMenuService_Factory() { return new AOMenuService(i0.ɵɵinject(i1.Platform), i0.ɵɵinject(i1.MenuController)); }, token: AOMenuService, providedIn: "root" });
    return AOMenuService;
}());
export { AOMenuService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AOMenuService.prototype._stacked;
    /** @type {?} */
    AOMenuService.prototype.platform;
    /** @type {?} */
    AOMenuService.prototype.controller;
}
