// ao-angular/esm5/lib/services/ao.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/ao.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { ActionSheetController, AlertController, LoadingController, ModalController, Platform, ToastController } from '@ionic/angular';
import { AOString } from '../utils/ao-string';
import { AOContainerService } from '../directives/ao-container/ao-container.service';
import { AOMenuService } from '../services/ao-menu.service';
import * as i0 from "@angular/core";
import * as i1 from "@ionic/angular";
import * as i2 from "./ao-menu.service";
import * as i3 from "../directives/ao-container/ao-container.service";
var AOService = /** @class */ (function () {
    function AOService(platform, ionLoading, ionAlert, ionModal, ionSheet, ionToast, menu, container) {
        this.platform = platform;
        this.ionLoading = ionLoading;
        this.ionAlert = ionAlert;
        this.ionModal = ionModal;
        this.ionSheet = ionSheet;
        this.ionToast = ionToast;
        this.menu = menu;
        this.container = container;
    }
    //--------------------------------------------------------------------------------------------------------------------
    // LOADING
    //--------------------------------------------------------------------------------------------------------------------
    //--------------------------------------------------------------------------------------------------------------------
    // LOADING
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} promise
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.wait = 
    //--------------------------------------------------------------------------------------------------------------------
    // LOADING
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} promise
     * @param {?=} opts
     * @return {?}
     */
    function (promise, opts) {
        if (opts === void 0) { opts = {}; }
        return this.ionLoading.create(tslib_1.__assign({ message: 'Carregando...' }, opts)).then((/**
         * @param {?} loading
         * @return {?}
         */
        function (loading) {
            return loading.present().then((/**
             * @return {?}
             */
            function () {
                promise.finally((/**
                 * @return {?}
                 */
                function () { return loading.dismiss(); }));
                return promise;
            }));
        }));
    };
    //--------------------------------------------------------------------------------------------------------------------
    // DROPDOWN
    //--------------------------------------------------------------------------------------------------------------------
    //--------------------------------------------------------------------------------------------------------------------
    // DROPDOWN
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?=} buttons
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.sheet = 
    //--------------------------------------------------------------------------------------------------------------------
    // DROPDOWN
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?=} buttons
     * @param {?=} opts
     * @return {?}
     */
    function (buttons, opts) {
        if (buttons === void 0) { buttons = {}; }
        if (opts === void 0) { opts = {}; }
        return this.ionSheet.create(tslib_1.__assign({ header: 'Opções...', buttons: buttons }, opts)).then((/**
         * @param {?} sheet
         * @return {?}
         */
        function (sheet) {
            return sheet.present().then((/**
             * @return {?}
             */
            function () {
                return sheet;
            }));
        }));
    };
    //--------------------------------------------------------------------------------------------------------------------
    // MODAL
    //--------------------------------------------------------------------------------------------------------------------
    //--------------------------------------------------------------------------------------------------------------------
    // MODAL
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} component
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.modal = 
    //--------------------------------------------------------------------------------------------------------------------
    // MODAL
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} component
     * @param {?=} opts
     * @return {?}
     */
    function (component, opts) {
        if (opts === void 0) { opts = {}; }
        return this.ionModal.create(tslib_1.__assign({ component: component }, opts)).then((/**
         * @param {?} modal
         * @return {?}
         */
        function (modal) {
            return modal.present().then((/**
             * @return {?}
             */
            function () {
                return modal;
            }));
        }));
    };
    //--------------------------------------------------------------------------------------------------------------------
    // ALERT
    //--------------------------------------------------------------------------------------------------------------------
    //--------------------------------------------------------------------------------------------------------------------
    // ALERT
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.alert = 
    //--------------------------------------------------------------------------------------------------------------------
    // ALERT
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    function (message, opts) {
        if (opts === void 0) { opts = {}; }
        return this.ionAlert.create(tslib_1.__assign({ header: 'Atenção...', message: message, buttons: [
                { text: 'OK', role: 'OK' }
            ] }, opts)).then((/**
         * @param {?} alert
         * @return {?}
         */
        function (alert) {
            return alert.present().then((/**
             * @return {?}
             */
            function () {
                return alert.onDidDismiss();
            }));
        }));
    };
    //--------------------------------------------------------------------------------------------------------------------
    // PROMPT
    //--------------------------------------------------------------------------------------------------------------------
    //--------------------------------------------------------------------------------------------------------------------
    // PROMPT
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} inputs
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.prompt = 
    //--------------------------------------------------------------------------------------------------------------------
    // PROMPT
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} inputs
     * @param {?=} opts
     * @return {?}
     */
    function (inputs, opts) {
        if (opts === void 0) { opts = {}; }
        return this.alert('', tslib_1.__assign({ header: 'Preencha...', inputs: inputs, buttons: [
                { text: 'CANCELAR', role: 'CANCEL' },
                { text: 'OK', role: 'OK' }
            ] }, opts));
    };
    //--------------------------------------------------------------------------------------------------------------------
    // CONFIRM
    //--------------------------------------------------------------------------------------------------------------------
    //--------------------------------------------------------------------------------------------------------------------
    // CONFIRM
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} message
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.confirm = 
    //--------------------------------------------------------------------------------------------------------------------
    // CONFIRM
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} message
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    function (message, yes, no, opts) {
        if (yes === void 0) { yes = (/**
         * @return {?}
         */
        function () { return null; }); }
        if (no === void 0) { no = (/**
         * @return {?}
         */
        function () { return null; }); }
        if (opts === void 0) { opts = {}; }
        return this.alert(message, tslib_1.__assign({ header: 'Confirmação...', buttons: [
                { text: 'NÃO', role: 'NO' },
                { text: 'SIM', role: 'YES' }
            ] }, opts)).then((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            response.role == 'YES' ? yes() : no();
            return response;
        }));
    };
    /**
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.confirmOperation = /**
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    function (yes, no, opts) {
        if (yes === void 0) { yes = (/**
         * @return {?}
         */
        function () { return null; }); }
        if (no === void 0) { no = (/**
         * @return {?}
         */
        function () { return null; }); }
        if (opts === void 0) { opts = {}; }
        return this.confirm('Deseja realmente <b>realizar</b> esta operação?', yes, no, opts);
    };
    /**
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.confirmActivate = /**
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    function (yes, no, opts) {
        if (yes === void 0) { yes = (/**
         * @return {?}
         */
        function () { return null; }); }
        if (no === void 0) { no = (/**
         * @return {?}
         */
        function () { return null; }); }
        if (opts === void 0) { opts = {}; }
        return this.confirm('Deseja realmente <b>ativar</b> este item?', yes, no, opts);
    };
    /**
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.confirmDeactivate = /**
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    function (yes, no, opts) {
        if (yes === void 0) { yes = (/**
         * @return {?}
         */
        function () { return null; }); }
        if (no === void 0) { no = (/**
         * @return {?}
         */
        function () { return null; }); }
        if (opts === void 0) { opts = {}; }
        return this.confirm('Deseja realmente <b>desativar</b> este item?', yes, no, opts);
    };
    /**
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.confirmDestroy = /**
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    function (yes, no, opts) {
        if (yes === void 0) { yes = (/**
         * @return {?}
         */
        function () { return null; }); }
        if (no === void 0) { no = (/**
         * @return {?}
         */
        function () { return null; }); }
        if (opts === void 0) { opts = {}; }
        return this.confirm('Deseja realmente <b>excluir</b> este item?', yes, no, opts);
    };
    //--------------------------------------------------------------------------------------------------------------------
    // TOAST
    //--------------------------------------------------------------------------------------------------------------------
    //--------------------------------------------------------------------------------------------------------------------
    // TOAST
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.toast = 
    //--------------------------------------------------------------------------------------------------------------------
    // TOAST
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    function (message, opts) {
        if (opts === void 0) { opts = {}; }
        return this.ionToast.create(tslib_1.__assign({ message: message, duration: AOString.readingTime(message) }, opts)).then((/**
         * @param {?} toast
         * @return {?}
         */
        function (toast) {
            return toast.present().then((/**
             * @return {?}
             */
            function () {
                return toast.onDidDismiss();
            }));
        }));
    };
    /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.info = /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    function (message, opts) {
        if (opts === void 0) { opts = {}; }
        //console.info(message);
        return this.toast(message, tslib_1.__assign({ color: 'tertiary' }, opts));
    };
    /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.success = /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    function (message, opts) {
        if (opts === void 0) { opts = {}; }
        //console.log(message);
        return this.toast(message, tslib_1.__assign({ color: 'success' }, opts));
    };
    /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.warning = /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    function (message, opts) {
        if (opts === void 0) { opts = {}; }
        console.warn(message);
        return this.toast(message, tslib_1.__assign({ color: 'warning' }, opts));
    };
    /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    AOService.prototype.danger = /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    function (message, opts) {
        if (opts === void 0) { opts = {}; }
        console.error(message, opts.error);
        return this.toast(message, tslib_1.__assign({ color: 'danger' }, opts));
    };
    AOService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    AOService.ctorParameters = function () { return [
        { type: Platform },
        { type: LoadingController },
        { type: AlertController },
        { type: ModalController },
        { type: ActionSheetController },
        { type: ToastController },
        { type: AOMenuService },
        { type: AOContainerService }
    ]; };
    /** @nocollapse */ AOService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function AOService_Factory() { return new AOService(i0.ɵɵinject(i1.Platform), i0.ɵɵinject(i1.LoadingController), i0.ɵɵinject(i1.AlertController), i0.ɵɵinject(i1.ModalController), i0.ɵɵinject(i1.ActionSheetController), i0.ɵɵinject(i1.ToastController), i0.ɵɵinject(i2.AOMenuService), i0.ɵɵinject(i3.AOContainerService)); }, token: AOService, providedIn: "root" });
    return AOService;
}());
export { AOService };
if (false) {
    /** @type {?} */
    AOService.prototype.platform;
    /** @type {?} */
    AOService.prototype.ionLoading;
    /** @type {?} */
    AOService.prototype.ionAlert;
    /** @type {?} */
    AOService.prototype.ionModal;
    /** @type {?} */
    AOService.prototype.ionSheet;
    /** @type {?} */
    AOService.prototype.ionToast;
    /** @type {?} */
    AOService.prototype.menu;
    /** @type {?} */
    AOService.prototype.container;
}
