// ao-angular/esm5/lib/helpers/ao-faunadb-pagination.helper.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/helpers/ao-faunadb-pagination.helper.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import * as FaunaDB from 'faunadb';
var AOFaunaDBPaginationHelper = /** @class */ (function () {
    function AOFaunaDBPaginationHelper(_ao, _db) {
        this._ao = _ao;
        this._db = _db;
        this._filters_show = true;
        this._filters_model = {};
        this._filters = {};
        this._size = 20;
    }
    //--------------------------------------------------------------------------------------------------------------------
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?=} filters
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.filters = 
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?=} filters
     * @return {?}
     */
    function (filters) {
        if (filters) {
            this._filters_model = tslib_1.__assign({}, filters);
            this._filters = tslib_1.__assign({}, filters);
            return this;
        }
        return this._filters;
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.openFilters = /**
     * @return {?}
     */
    function () {
        this._filters_show = true;
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.closeFilters = /**
     * @return {?}
     */
    function () {
        this._filters_show = false;
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.toggleFilters = /**
     * @return {?}
     */
    function () {
        this._filters_show = !this._filters_show;
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.showFilters = /**
     * @return {?}
     */
    function () {
        return this._filters_show;
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.hideFilters = /**
     * @return {?}
     */
    function () {
        return !this._filters_show;
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.hasFilters = /**
     * @return {?}
     */
    function () {
        var _this = this;
        return Object
            .keys(this._filters_model)
            .filter((/**
         * @param {?} key
         * @return {?}
         */
        function (key) { return _this._filters_model[key] !== _this._filters[key]; }))
            .length > 0;
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.resetFilters = /**
     * @return {?}
     */
    function () {
        this._filters = tslib_1.__assign({}, this._filters_model);
        this.openFilters();
    };
    //--------------------------------------------------------------------------------------------------------------------
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} columns
     * @return {THIS}
     */
    AOFaunaDBPaginationHelper.prototype.select = 
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} columns
     * @return {THIS}
     */
    function (columns) {
        return ((/** @type {?} */ (this))._columns = tslib_1.__spread(columns)) && (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} query
     * @return {THIS}
     */
    AOFaunaDBPaginationHelper.prototype.query = /**
     * @template THIS
     * @this {THIS}
     * @param {?} query
     * @return {THIS}
     */
    function (query) {
        return ((/** @type {?} */ (this))._query = query) && (/** @type {?} */ (this));
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.search = /**
     * @return {?}
     */
    function () {
        this._last_after = undefined;
        this._refresh = undefined;
        this._search = this._query(FaunaDB.query, tslib_1.__assign({}, this.filters()));
        return this.first();
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.first = /**
     * @return {?}
     */
    function () {
        return this.paginate(this._search, {
            size: this._size,
            before: undefined,
            after: undefined,
        });
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.previous = /**
     * @return {?}
     */
    function () {
        return this.paginate(this._search, {
            size: this._size,
            after: undefined,
            before: this._response.before,
        });
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.next = /**
     * @return {?}
     */
    function () {
        return this.paginate(this._search, {
            size: this._size,
            after: this._response.after,
            before: undefined,
        });
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.last = /**
     * @return {?}
     */
    function () {
        return this.paginate(this._search, {
            size: this._size,
            after: this._last_after,
            before: undefined,
        });
    };
    /**
     * @param {?=} event
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.refresh = /**
     * @param {?=} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var promise;
        if (this.empty()) {
            if (this._response) {
                promise = this.previous();
            }
            else {
                promise = this.search();
            }
        }
        else {
            promise = this.paginate(this._search, this._refresh);
        }
        return promise.then((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            event ? event.target.complete() : null;
            return response;
        }));
    };
    /**
     * @private
     * @param {?} set
     * @param {?=} opts
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.paginate = /**
     * @private
     * @param {?} set
     * @param {?=} opts
     * @return {?}
     */
    function (set, opts) {
        var _this = this;
        this._refresh = tslib_1.__assign({}, opts);
        /** @type {?} */
        var paginate = FaunaDB.query.Paginate(set, opts);
        if (this._columns) {
            paginate = FaunaDB.query.Map(paginate, (/**
             * @param {?} item
             * @return {?}
             */
            function (item) { return _this._columns.reduce((/**
             * @param {?} select
             * @param {?} column
             * @param {?} index
             * @return {?}
             */
            function (select, column, index) { return (select[column] = FaunaDB.query.Select(index + 1, item)) && select; }), {}); }));
        }
        return this._db.query((/**
         * @return {?}
         */
        function () { return paginate; }))
            .then((/**
         * @param {?} response
         * @return {?}
         */
        function (response) {
            if (_this._response && _this._response.after && !response.after) {
                _this._last_after = _this._response.after;
            }
            _this._response = response;
            _this.closeFilters();
        })).catch((/**
         * @param {?} error
         * @return {?}
         */
        function (error) {
            _this._ao.danger('Oops! Falha ao realizar a buscar :(', { error: error });
        }));
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.hasPrevious = /**
     * @return {?}
     */
    function () {
        return !!(this._response && this._response.before);
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.hasNext = /**
     * @return {?}
     */
    function () {
        return !!(this._response && this._response.after);
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.hasLast = /**
     * @return {?}
     */
    function () {
        return !!this._last_after;
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.isFirst = /**
     * @return {?}
     */
    function () {
        return !this.hasPrevious();
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.isLast = /**
     * @return {?}
     */
    function () {
        return !this.hasNext();
    };
    //--------------------------------------------------------------------------------------------------------------------
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.response = 
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @return {?}
     */
    function () {
        return this._response ? this._response : null;
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.items = /**
     * @return {?}
     */
    function () {
        return this._response && this._response.data ? this._response.data : [];
    };
    /**
     * @return {?}
     */
    AOFaunaDBPaginationHelper.prototype.empty = /**
     * @return {?}
     */
    function () {
        return this.items().length === 0;
    };
    return AOFaunaDBPaginationHelper;
}());
export { AOFaunaDBPaginationHelper };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AOFaunaDBPaginationHelper.prototype._filters_show;
    /**
     * @type {?}
     * @private
     */
    AOFaunaDBPaginationHelper.prototype._filters_model;
    /**
     * @type {?}
     * @private
     */
    AOFaunaDBPaginationHelper.prototype._filters;
    /**
     * @type {?}
     * @private
     */
    AOFaunaDBPaginationHelper.prototype._query;
    /**
     * @type {?}
     * @private
     */
    AOFaunaDBPaginationHelper.prototype._columns;
    /**
     * @type {?}
     * @private
     */
    AOFaunaDBPaginationHelper.prototype._search;
    /**
     * @type {?}
     * @private
     */
    AOFaunaDBPaginationHelper.prototype._size;
    /**
     * @type {?}
     * @private
     */
    AOFaunaDBPaginationHelper.prototype._response;
    /**
     * @type {?}
     * @private
     */
    AOFaunaDBPaginationHelper.prototype._last_after;
    /**
     * @type {?}
     * @private
     */
    AOFaunaDBPaginationHelper.prototype._refresh;
    /**
     * @type {?}
     * @private
     */
    AOFaunaDBPaginationHelper.prototype._ao;
    /**
     * @type {?}
     * @private
     */
    AOFaunaDBPaginationHelper.prototype._db;
}
