// ao-angular/esm2015/lib/directives/ao-container/ao-container.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/ao-container/ao-container.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ComponentFactoryResolver, Directive, ViewContainerRef } from '@angular/core';
export class AOContainerDirective {
    /**
     * @param {?} _content
     * @param {?} _resolver
     */
    constructor(_content, _resolver) {
        this._content = _content;
        this._resolver = _resolver;
        this._component = null;
    }
    /**
     * @return {?}
     */
    component() {
        return this._component;
    }
    /**
     * @return {?}
     */
    content() {
        return this._content;
    }
    /**
     * @return {?}
     */
    clear() {
        this.content().clear();
    }
    /**
     * @param {?} component
     * @return {?}
     */
    open(component) {
        this.clear();
        this._component = this.content().createComponent(typeof component !== 'function' ? component : this._resolver.resolveComponentFactory(component));
        this._component.instance.directive = this;
        return this._component.instance;
    }
}
AOContainerDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ao-container]'
            },] }
];
/** @nocollapse */
AOContainerDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: ComponentFactoryResolver }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    AOContainerDirective.prototype._component;
    /**
     * @type {?}
     * @private
     */
    AOContainerDirective.prototype._content;
    /**
     * @type {?}
     * @private
     */
    AOContainerDirective.prototype._resolver;
}
