// ao-angular/esm2015/lib/directives/ao-container/ao-container.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/ao-container/ao-container.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class AOContainerService {
    constructor() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} directive
     * @return {THIS}
     */
    init(directive) {
        (/** @type {?} */ (this))._directive = directive;
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} component
     * @param {?=} data
     * @return {THIS}
     */
    open(component, data = {}) {
        /** @type {?} */
        const instance = (/** @type {?} */ (this))._directive.open(component);
        instance.data = data;
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    close() {
        (/** @type {?} */ (this))._directive.clear();
        return (/** @type {?} */ (this));
    }
}
AOContainerService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
AOContainerService.ctorParameters = () => [];
/** @nocollapse */ AOContainerService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function AOContainerService_Factory() { return new AOContainerService(); }, token: AOContainerService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    AOContainerService.prototype._directive;
}
