// ao-angular/esm2015/lib/utils/ao-fake.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/ao-fake.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// @dynamic
export class AOFake {
    /**
     * @param {?} value
     * @param {?=} decimals
     * @return {?}
     */
    static floorDecimal(value, decimals = 2) {
        return Math.floor(value * Math.pow(10, decimals)) / Math.pow(10, decimals);
    }
    /**
     * @param {?=} min
     * @param {?=} max
     * @param {?=} decimals
     * @return {?}
     */
    static randomFloat(min = 0, max = 100, decimals = 2) {
        return Math.floor((min + (Math.random() * (max - min))) * Math.pow(10, decimals)) / Math.pow(10, decimals);
    }
    /**
     * @param {?=} min
     * @param {?=} max
     * @return {?}
     */
    static randomInteger(min = 0, max = 100) {
        return Math.floor((min + (Math.random() * (max - min))) * Math.pow(10, 0)) / Math.pow(10, 0);
    }
    /**
     * @param {?=} chars
     * @return {?}
     */
    static randomChar(chars = 'qwertyuiopasdfghjklzxcvbnm1234567890') {
        return chars.charAt(this.randomInteger(0, chars.length - 1));
    }
    /**
     * @param {?=} min
     * @param {?=} max
     * @return {?}
     */
    static randomWord(min = 1, max = 10) {
        /** @type {?} */
        const size = this.randomInteger(min, max);
        /** @type {?} */
        let word = '';
        while (word.length < size) {
            word += this.randomChar();
        }
        return word;
    }
    /**
     * @param {?=} min
     * @param {?=} max
     * @return {?}
     */
    static randomWords(min = 1, max = 10) {
        /** @type {?} */
        const qt = this.randomInteger(min, max);
        /** @type {?} */
        const words = [];
        while (words.length < qt) {
            words.push(this.randomWord());
        }
        return words.join(' ');
    }
    /**
     * @param {?=} fator
     * @return {?}
     */
    static randomBoolean(fator = 50) {
        return this.randomInteger() <= fator;
    }
    /**
     * @param {?} elements
     * @return {?}
     */
    static randomElement(elements) {
        return elements[this.randomInteger(0, elements.length - 1)];
    }
    /**
     * @param {?} elements
     * @param {?=} min
     * @param {?=} max
     * @return {?}
     */
    static randomElements(elements, min = 1, max = null) {
        min = min && min > 1 ? Math.floor(min) : 1;
        max = max && max < elements.length ? Math.floor(max) : this.randomInteger(min, elements.length - 1);
        /** @type {?} */
        const qt = this.randomInteger(min, max);
        /** @type {?} */
        const selection = [];
        while (selection.length < qt) {
            selection.push(this.randomElement(elements));
        }
        return selection;
    }
}
