// ao-angular/esm2015/lib/utils/ao-validators.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/utils/ao-validators.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { AOValidations } from './ao-validations';
// @dynamic
export class AOValidators {
    /**
     * @param {?} control
     * @return {?}
     */
    static touched(control) {
        if (control.pristine) {
            return true;
        }
        control.markAsTouched();
        return false;
    }
    /**
     * @param {?} callback
     * @return {?}
     */
    static custom(callback) {
        return (/**
         * @param {?} control
         * @return {?}
         */
        control => {
            return AOValidators.touched(control) || callback(control.value) ? null : { invalidCustom: true };
        });
    }
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} control
     * @return {?}
     */
    static email(control) {
        return AOValidators.touched(control) || AOValidations.isEmail(control.value) ? null : { invalidEmail: true };
    }
    /**
     * @param {?} control
     * @return {?}
     */
    static url(control) {
        return AOValidators.touched(control) || AOValidations.isUrl(control.value) ? null : { invalidUrl: true };
    }
    /**
     * @param {?} control
     * @return {?}
     */
    static phone(control) {
        return AOValidators.touched(control) || AOValidations.isPhone(control.value) ? null : { invalidPhone: true };
    }
    /**
     * @param {?} control
     * @return {?}
     */
    static phoneINTER(control) {
        return AOValidators.touched(control) || AOValidations.isPhoneINTER(control.value) ? null : { invalidPhone: true };
    }
    /**
     * @param {?} control
     * @return {?}
     */
    static zipCode(control) {
        return AOValidators.touched(control) || AOValidations.isZipCode(control.value) ? null : { invalidZip: true };
    }
    /**
     * @param {?} control
     * @return {?}
     */
    static zipCodeINTER(control) {
        return AOValidators.touched(control) || AOValidations.isZipCodeINTER(control.value) ? null : { invalidZip: true };
    }
    /**
     * @param {?} control
     * @return {?}
     */
    static cpf(control) {
        return AOValidators.touched(control) || AOValidations.isCpf(control.value) ? null : { invalidZip: true };
    }
    /**
     * @param {?} control
     * @return {?}
     */
    static cnpj(control) {
        return AOValidators.touched(control) || AOValidations.isCnpj(control.value) ? null : { invalidZip: true };
    }
    /**
     * @param {?} control
     * @return {?}
     */
    static cpfCnpj(control) {
        return AOValidators.touched(control) || AOValidations.isCpfCnpj(control.value) ? null : { invalidZip: true };
    }
    /**
     * @param {?} control
     * @return {?}
     */
    static float(control) {
        return AOValidators.touched(control) || AOValidations.isFloat(control.value) ? null : { invalidNumber: true };
    }
    /**
     * @param {?} control
     * @return {?}
     */
    static floatSigned(control) {
        return AOValidators.touched(control) || AOValidations.isFloat_S(control.value) ? null : { invalidNumber: true };
    }
    /**
     * @param {?} control
     * @return {?}
     */
    static floatUnsigned(control) {
        return AOValidators.touched(control) || AOValidations.isFloat_U(control.value) ? null : { invalidNumber: true };
    }
    /**
     * @param {?} control
     * @return {?}
     */
    static int(control) {
        return AOValidators.touched(control) || AOValidations.isInt(control.value) ? null : { invalidNumber: true };
    }
    /**
     * @param {?} control
     * @return {?}
     */
    static intSigned(control) {
        return AOValidators.touched(control) || AOValidations.isInt_S(control.value) ? null : { invalidNumber: true };
    }
    /**
     * @param {?} control
     * @return {?}
     */
    static intUnsigned(control) {
        return AOValidators.touched(control) || AOValidations.isInt_U(control.value) ? null : { invalidNumber: true };
    }
}
