// ao-angular/esm2015/lib/services/ao-firebase-connection.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/ao-firebase-connection.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { AngularFireAuth } from '@angular/fire/auth';
import { AngularFireDatabase } from '@angular/fire/database';
import { AngularFireStorage } from '@angular/fire/storage';
import { AngularFirestore } from '@angular/fire/firestore';
import { AngularFireMessaging } from '@angular/fire/messaging';
export class AOFirebaseConnectionService {
    /**
     * @param {?} _name
     * @param {?} _hash
     * @param {?} _app
     * @param {?} _zone
     */
    constructor(_name, _hash, _app, _zone) {
        this._name = _name;
        this._hash = _hash;
        this._app = _app;
        this._zone = _zone;
        this._configs = {};
    }
    /**
     * @return {?}
     */
    app() {
        return this._app;
    }
    /**
     * @return {?}
     */
    name() {
        return this._name;
    }
    /**
     * @return {?}
     */
    hash() {
        return this._hash;
    }
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @return {?}
     */
    auth() {
        return this._auth
            ? this._auth
            : this._auth = new AngularFireAuth(this._configs, this._name, null, this._zone);
    }
    /**
     * @return {?}
     */
    database() {
        return this._database
            ? this._database
            : this._database = new AngularFireDatabase(this._configs, this._name, null, null, this._zone);
    }
    /**
     * @return {?}
     */
    firestore() {
        return this._firestore
            ? this._firestore
            : this._firestore = new AngularFirestore(this._configs, this._name, null, null, null, this._zone, null);
    }
    /**
     * @return {?}
     */
    messaging() {
        return this._messaging
            ? this._messaging
            : this._messaging = new AngularFireMessaging(this._configs, this._name, null, this._zone);
    }
    /**
     * @return {?}
     */
    storage() {
        return this._storage
            ? this._storage
            : this._storage = new AngularFireStorage(this._configs, this._name, null, null, this._zone);
    }
    // -------------------------------------------------------------------------------------------------------------------
    /**
     * @return {?}
     */
    logout() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this._auth) {
                try {
                    yield this._auth.auth.signOut();
                }
                catch (e) {
                    console.warn('Oops! Failed to run signOut.');
                }
            }
            return Promise.resolve(true);
        });
    }
    /**
     * @return {?}
     */
    destroy() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            yield this.logout();
            yield this._app.delete();
            return Promise.resolve(true);
        });
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    AOFirebaseConnectionService.prototype._configs;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseConnectionService.prototype._auth;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseConnectionService.prototype._database;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseConnectionService.prototype._firestore;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseConnectionService.prototype._messaging;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseConnectionService.prototype._storage;
    /**
     * @type {?}
     * @protected
     */
    AOFirebaseConnectionService.prototype._name;
    /**
     * @type {?}
     * @protected
     */
    AOFirebaseConnectionService.prototype._hash;
    /**
     * @type {?}
     * @protected
     */
    AOFirebaseConnectionService.prototype._app;
    /**
     * @type {?}
     * @protected
     */
    AOFirebaseConnectionService.prototype._zone;
}
