// ao-angular/esm2015/lib/services/ao-firebase-factory.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/ao-firebase-factory.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Inject, Injectable, NgZone } from '@angular/core';
import { FirebaseApp } from '@angular/fire';
import { _firebaseAppFactory } from '@angular/fire/firebase.app.module';
import { AOFirebaseConnectionService } from './ao-firebase-connection.service';
import { AOEnvService } from './ao-env.service';
import * as i0 from "@angular/core";
import * as i1 from "@angular/fire";
import * as i2 from "./ao-env.service";
export class AOFirebaseFactoryService {
    /**
     * @param {?} _app
     * @param {?} _zone
     * @param {?} _environment
     */
    constructor(_app, _zone, _environment) {
        this._app = _app;
        this._zone = _zone;
        this._environment = _environment;
        this.connections = {};
        this.factory(this._environment.firebase.name, this._environment.firebase.configs, this._app);
    }
    /**
     * @param {?=} name
     * @return {?}
     */
    conn(name) {
        if (this.connections[name]) {
            return this.connections[name];
        }
        throw new Error('FirebaseApp(' + name + ') not found.');
    }
    /**
     * @param {?} configs
     * @return {?}
     */
    makeHash(configs) {
        return JSON.stringify({
            projectId: configs.projectId,
            apiKey: configs.apiKey
        });
    }
    /**
     * @param {?} name
     * @param {?=} configs
     * @param {?=} app
     * @return {?}
     */
    factory(name, configs, app) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            /** @type {?} */
            const hash = this.makeHash(configs);
            if (this.connections[name]) {
                if (this.connections[name].hash() === hash) {
                    return Promise.resolve(true);
                }
                else {
                    yield this.connections[name].destroy();
                }
            }
            if (!app) {
                app = _firebaseAppFactory((/** @type {?} */ (configs)), this._zone, name);
            }
            this.connections[name] = new AOFirebaseConnectionService(name, hash, app, this._zone);
            return Promise.resolve(true);
        });
    }
    /**
     * @param {?=} only
     * @param {?=} except
     * @return {?}
     */
    destroy(only = [], except = ['master']) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (only.length === 0) {
                only = Object.keys(this.connections);
            }
            /** @type {?} */
            const promises = [];
            Object.keys(this.connections).forEach((/**
             * @param {?} name
             * @return {?}
             */
            name => {
                if (only.indexOf(name) >= 0 && except.indexOf(name) === -1) {
                    promises.push(this.conn(name).destroy());
                }
            }));
            return Promise.all(promises).then((/**
             * @return {?}
             */
            () => true));
        });
    }
}
AOFirebaseFactoryService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
AOFirebaseFactoryService.ctorParameters = () => [
    { type: FirebaseApp },
    { type: NgZone },
    { type: undefined, decorators: [{ type: Inject, args: [AOEnvService,] }] }
];
/** @nocollapse */ AOFirebaseFactoryService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function AOFirebaseFactoryService_Factory() { return new AOFirebaseFactoryService(i0.ɵɵinject(i1.FirebaseApp), i0.ɵɵinject(i0.NgZone), i0.ɵɵinject(i2.AOEnvService)); }, token: AOFirebaseFactoryService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    AOFirebaseFactoryService.prototype.connections;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseFactoryService.prototype._app;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseFactoryService.prototype._zone;
    /**
     * @type {?}
     * @private
     */
    AOFirebaseFactoryService.prototype._environment;
}
