// ao-angular/esm2015/lib/services/ao-menu.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/ao-menu.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { MenuController, Platform } from '@ionic/angular';
import * as i0 from "@angular/core";
import * as i1 from "@ionic/angular";
export class AOMenuService {
    /**
     * @param {?} platform
     * @param {?} controller
     */
    constructor(platform, controller) {
        this.platform = platform;
        this.controller = controller;
        this._stacked = ['lg'];
    }
    /**
     * @return {?}
     */
    disable() {
        return this.controller.enable(false, 'AO_APP_MENU');
    }
    /**
     * @return {?}
     */
    enable() {
        return this.controller.enable(true, 'AO_APP_MENU');
    }
    /**
     * @return {?}
     */
    stackable() {
        return this.platform.width() > 991;
    }
    /**
     * @return {?}
     */
    stacked() {
        return this._stacked;
    }
    /**
     * @return {?}
     */
    stackToggle() {
        this._stacked = this._stacked == false ? ['lg'] : false;
    }
}
AOMenuService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
AOMenuService.ctorParameters = () => [
    { type: Platform },
    { type: MenuController }
];
/** @nocollapse */ AOMenuService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function AOMenuService_Factory() { return new AOMenuService(i0.ɵɵinject(i1.Platform), i0.ɵɵinject(i1.MenuController)); }, token: AOMenuService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    AOMenuService.prototype._stacked;
    /** @type {?} */
    AOMenuService.prototype.platform;
    /** @type {?} */
    AOMenuService.prototype.controller;
}
