// ao-angular/esm2015/lib/services/ao.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/services/ao.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ActionSheetController, AlertController, LoadingController, ModalController, Platform, ToastController } from '@ionic/angular';
import { AOString } from '../utils/ao-string';
import { AOContainerService } from '../directives/ao-container/ao-container.service';
import { AOMenuService } from '../services/ao-menu.service';
import * as i0 from "@angular/core";
import * as i1 from "@ionic/angular";
import * as i2 from "./ao-menu.service";
import * as i3 from "../directives/ao-container/ao-container.service";
export class AOService {
    /**
     * @param {?} platform
     * @param {?} ionLoading
     * @param {?} ionAlert
     * @param {?} ionModal
     * @param {?} ionSheet
     * @param {?} ionToast
     * @param {?} menu
     * @param {?} container
     */
    constructor(platform, ionLoading, ionAlert, ionModal, ionSheet, ionToast, menu, container) {
        this.platform = platform;
        this.ionLoading = ionLoading;
        this.ionAlert = ionAlert;
        this.ionModal = ionModal;
        this.ionSheet = ionSheet;
        this.ionToast = ionToast;
        this.menu = menu;
        this.container = container;
    }
    //--------------------------------------------------------------------------------------------------------------------
    // LOADING
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} promise
     * @param {?=} opts
     * @return {?}
     */
    wait(promise, opts = {}) {
        return this.ionLoading.create(Object.assign({ message: 'Carregando...' }, opts)).then((/**
         * @param {?} loading
         * @return {?}
         */
        loading => {
            return loading.present().then((/**
             * @return {?}
             */
            () => {
                promise.finally((/**
                 * @return {?}
                 */
                () => loading.dismiss()));
                return promise;
            }));
        }));
    }
    //--------------------------------------------------------------------------------------------------------------------
    // DROPDOWN
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?=} buttons
     * @param {?=} opts
     * @return {?}
     */
    sheet(buttons = {}, opts = {}) {
        return this.ionSheet.create(Object.assign({ header: 'Opções...', buttons }, opts)).then((/**
         * @param {?} sheet
         * @return {?}
         */
        sheet => {
            return sheet.present().then((/**
             * @return {?}
             */
            () => {
                return sheet;
            }));
        }));
    }
    //--------------------------------------------------------------------------------------------------------------------
    // MODAL
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} component
     * @param {?=} opts
     * @return {?}
     */
    modal(component, opts = {}) {
        return this.ionModal.create(Object.assign({ component }, opts)).then((/**
         * @param {?} modal
         * @return {?}
         */
        modal => {
            return modal.present().then((/**
             * @return {?}
             */
            () => {
                return modal;
            }));
        }));
    }
    //--------------------------------------------------------------------------------------------------------------------
    // ALERT
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    alert(message, opts = {}) {
        return this.ionAlert.create(Object.assign({ header: 'Atenção...', message, buttons: [
                { text: 'OK', role: 'OK' }
            ] }, opts)).then((/**
         * @param {?} alert
         * @return {?}
         */
        alert => {
            return alert.present().then((/**
             * @return {?}
             */
            () => {
                return alert.onDidDismiss();
            }));
        }));
    }
    //--------------------------------------------------------------------------------------------------------------------
    // PROMPT
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} inputs
     * @param {?=} opts
     * @return {?}
     */
    prompt(inputs, opts = {}) {
        return this.alert('', Object.assign({ header: 'Preencha...', inputs, buttons: [
                { text: 'CANCELAR', role: 'CANCEL' },
                { text: 'OK', role: 'OK' }
            ] }, opts));
    }
    //--------------------------------------------------------------------------------------------------------------------
    // CONFIRM
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} message
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    confirm(message, yes = (/**
     * @return {?}
     */
    () => null), no = (/**
     * @return {?}
     */
    () => null), opts = {}) {
        return this.alert(message, Object.assign({ header: 'Confirmação...', buttons: [
                { text: 'NÃO', role: 'NO' },
                { text: 'SIM', role: 'YES' }
            ] }, opts)).then((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            response.role == 'YES' ? yes() : no();
            return response;
        }));
    }
    /**
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    confirmOperation(yes = (/**
     * @return {?}
     */
    () => null), no = (/**
     * @return {?}
     */
    () => null), opts = {}) {
        return this.confirm('Deseja realmente <b>realizar</b> esta operação?', yes, no, opts);
    }
    /**
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    confirmActivate(yes = (/**
     * @return {?}
     */
    () => null), no = (/**
     * @return {?}
     */
    () => null), opts = {}) {
        return this.confirm('Deseja realmente <b>ativar</b> este item?', yes, no, opts);
    }
    /**
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    confirmDeactivate(yes = (/**
     * @return {?}
     */
    () => null), no = (/**
     * @return {?}
     */
    () => null), opts = {}) {
        return this.confirm('Deseja realmente <b>desativar</b> este item?', yes, no, opts);
    }
    /**
     * @param {?=} yes
     * @param {?=} no
     * @param {?=} opts
     * @return {?}
     */
    confirmDestroy(yes = (/**
     * @return {?}
     */
    () => null), no = (/**
     * @return {?}
     */
    () => null), opts = {}) {
        return this.confirm('Deseja realmente <b>excluir</b> este item?', yes, no, opts);
    }
    //--------------------------------------------------------------------------------------------------------------------
    // TOAST
    //--------------------------------------------------------------------------------------------------------------------
    /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    toast(message, opts = {}) {
        return this.ionToast.create(Object.assign({ message, duration: AOString.readingTime(message) }, opts)).then((/**
         * @param {?} toast
         * @return {?}
         */
        toast => {
            return toast.present().then((/**
             * @return {?}
             */
            () => {
                return toast.onDidDismiss();
            }));
        }));
    }
    /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    info(message, opts = {}) {
        //console.info(message);
        return this.toast(message, Object.assign({ color: 'tertiary' }, opts));
    }
    /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    success(message, opts = {}) {
        //console.log(message);
        return this.toast(message, Object.assign({ color: 'success' }, opts));
    }
    /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    warning(message, opts = {}) {
        console.warn(message);
        return this.toast(message, Object.assign({ color: 'warning' }, opts));
    }
    /**
     * @param {?} message
     * @param {?=} opts
     * @return {?}
     */
    danger(message, opts = {}) {
        console.error(message, opts.error);
        return this.toast(message, Object.assign({ color: 'danger' }, opts));
    }
}
AOService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
AOService.ctorParameters = () => [
    { type: Platform },
    { type: LoadingController },
    { type: AlertController },
    { type: ModalController },
    { type: ActionSheetController },
    { type: ToastController },
    { type: AOMenuService },
    { type: AOContainerService }
];
/** @nocollapse */ AOService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function AOService_Factory() { return new AOService(i0.ɵɵinject(i1.Platform), i0.ɵɵinject(i1.LoadingController), i0.ɵɵinject(i1.AlertController), i0.ɵɵinject(i1.ModalController), i0.ɵɵinject(i1.ActionSheetController), i0.ɵɵinject(i1.ToastController), i0.ɵɵinject(i2.AOMenuService), i0.ɵɵinject(i3.AOContainerService)); }, token: AOService, providedIn: "root" });
if (false) {
    /** @type {?} */
    AOService.prototype.platform;
    /** @type {?} */
    AOService.prototype.ionLoading;
    /** @type {?} */
    AOService.prototype.ionAlert;
    /** @type {?} */
    AOService.prototype.ionModal;
    /** @type {?} */
    AOService.prototype.ionSheet;
    /** @type {?} */
    AOService.prototype.ionToast;
    /** @type {?} */
    AOService.prototype.menu;
    /** @type {?} */
    AOService.prototype.container;
}
