// ao-angular/esm5/lib/components/ao-app/ao-app.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/components/ao-app/ao-app.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { AOService } from '../../services/ao.service';
var AOAppComponent = /** @class */ (function () {
    function AOAppComponent(ao) {
        this.ao = ao;
    }
    /**
     * @return {?}
     */
    AOAppComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    AOAppComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ao-app',
                    template: "<ion-app>\n  <ion-split-pane contentId=\"AO_APP_CONTENT\" [when]=\"ao.menu.stacked()\">\n\n    <ion-menu menuId=\"AO_APP_MENU\" contentId=\"AO_APP_CONTENT\" type=\"overlay\">\n      <ion-header>\n\n        <ng-content select=\"[slot=header]\"></ng-content>\n\n      </ion-header>\n      <ion-content>\n\n        <ng-content select=\"[slot=menu]\"></ng-content>\n\n      </ion-content>\n      <ion-header>\n\n        <ng-content select=\"[slot=footer]\"></ng-content>\n\n      </ion-header>\n    </ion-menu>\n\n    <ion-router-outlet id=\"AO_APP_CONTENT\"></ion-router-outlet>\n\n  </ion-split-pane>\n</ion-app>\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    AOAppComponent.ctorParameters = function () { return [
        { type: AOService }
    ]; };
    return AOAppComponent;
}());
export { AOAppComponent };
if (false) {
    /** @type {?} */
    AOAppComponent.prototype.ao;
}
