// ao-angular/esm5/lib/directives/ao-container/ao-container.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/ao-container/ao-container.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, ViewChild, ViewContainerRef } from '@angular/core';
import { createAnimation } from '@ionic/core';
import { AOContainerDirective } from './ao-container.directive';
var AOContainerComponent = /** @class */ (function () {
    function AOContainerComponent(_content) {
        this._content = _content;
    }
    /**
     * @return {?}
     */
    AOContainerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var animation = createAnimation()
            // .addElement(document.querySelector('section'))
            .addElement(this.tref.nativeElement.innerHTML)
            // .addElement(this.tref.)
            .easing("ease-in-out")
            .duration(1000)
            .direction("alternate")
            .iterations(Infinity)
            .fromTo('opacity', '0.0', '0.8')
            .keyframes([
            { offset: 0, transform: "scale(1)", opacity: "1" },
            { offset: 1, transform: "scale(1.5)", opacity: "0.5" }
        ]);
        animation.play();
    };
    /**
     * @return {?}
     */
    AOContainerComponent.prototype.close = /**
     * @return {?}
     */
    function () {
        this.directive.clear();
    };
    AOContainerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ao-container',
                    template: "<div #tref>\r\n  <ion-backdrop (ionBackdropTap)=\"close()\" class=\"menu-type-overlay show-backdrop\"></ion-backdrop>\r\n\r\n  <section class=\"down\">\r\n\r\n    <div class=\"ion-page\">\r\n      <ion-header>\r\n        <ion-toolbar>\r\n          <ion-title>AOContainerComponent</ion-title>\r\n        </ion-toolbar>\r\n      </ion-header>\r\n      <ion-content>\r\n        <ion-button expand=\"block\" (click)=\"close()\">CLOSE</ion-button>\r\n      </ion-content>\r\n      <ion-footer>\r\n        <ion-toolbar>\r\n          <ion-title>AOContainerComponent</ion-title>\r\n        </ion-toolbar>\r\n      </ion-footer>\r\n    </div>\r\n\r\n  </section>\r\n</div>",
                    styles: ["section{position:absolute;width:300px;height:400px;z-index:2}section.top{top:0;left:50%;margin-left:-150px}section.down{bottom:0;left:50%;margin-left:-150px}section.right{right:0;top:50%;margin-top:-200px}section.left{left:0;top:50%;margin-top:-200px}"]
                }] }
    ];
    /** @nocollapse */
    AOContainerComponent.ctorParameters = function () { return [
        { type: ViewContainerRef }
    ]; };
    AOContainerComponent.propDecorators = {
        directive: [{ type: Input }],
        tref: [{ type: ViewChild, args: ["tref", { read: ElementRef, static: true },] }]
    };
    return AOContainerComponent;
}());
export { AOContainerComponent };
if (false) {
    /** @type {?} */
    AOContainerComponent.prototype.directive;
    /** @type {?} */
    AOContainerComponent.prototype.tref;
    /**
     * @type {?}
     * @private
     */
    AOContainerComponent.prototype._content;
}
