// ao-angular/esm5/lib/directives/ao-container/ao-container.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/ao-container/ao-container.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
var AOContainerService = /** @class */ (function () {
    function AOContainerService() {
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} directive
     * @return {THIS}
     */
    AOContainerService.prototype.init = /**
     * @template THIS
     * @this {THIS}
     * @param {?} directive
     * @return {THIS}
     */
    function (directive) {
        (/** @type {?} */ (this))._directive = directive;
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} component
     * @param {?=} data
     * @return {THIS}
     */
    AOContainerService.prototype.open = /**
     * @template THIS
     * @this {THIS}
     * @param {?} component
     * @param {?=} data
     * @return {THIS}
     */
    function (component, data) {
        if (data === void 0) { data = {}; }
        /** @type {?} */
        var instance = (/** @type {?} */ (this))._directive.open(component);
        instance.data = data;
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    AOContainerService.prototype.close = /**
     * @template THIS
     * @this {THIS}
     * @return {THIS}
     */
    function () {
        (/** @type {?} */ (this))._directive.clear();
        return (/** @type {?} */ (this));
    };
    AOContainerService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    AOContainerService.ctorParameters = function () { return []; };
    /** @nocollapse */ AOContainerService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function AOContainerService_Factory() { return new AOContainerService(); }, token: AOContainerService, providedIn: "root" });
    return AOContainerService;
}());
export { AOContainerService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AOContainerService.prototype._directive;
}
