// ao-angular/esm5/lib/directives/ao-container/ao-container.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/directives/ao-container/ao-container.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ComponentFactoryResolver, Directive, ViewContainerRef } from '@angular/core';
var AOContainerDirective = /** @class */ (function () {
    function AOContainerDirective(_content, _resolver) {
        this._content = _content;
        this._resolver = _resolver;
        this._component = null;
    }
    /**
     * @return {?}
     */
    AOContainerDirective.prototype.component = /**
     * @return {?}
     */
    function () {
        return this._component;
    };
    /**
     * @return {?}
     */
    AOContainerDirective.prototype.content = /**
     * @return {?}
     */
    function () {
        return this._content;
    };
    /**
     * @return {?}
     */
    AOContainerDirective.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.content().clear();
    };
    /**
     * @param {?} component
     * @return {?}
     */
    AOContainerDirective.prototype.open = /**
     * @param {?} component
     * @return {?}
     */
    function (component) {
        this.clear();
        this._component = this.content().createComponent(typeof component !== 'function' ? component : this._resolver.resolveComponentFactory(component));
        this._component.instance.directive = this;
        return this._component.instance;
    };
    AOContainerDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[ao-container]'
                },] }
    ];
    /** @nocollapse */
    AOContainerDirective.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: ComponentFactoryResolver }
    ]; };
    return AOContainerDirective;
}());
export { AOContainerDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    AOContainerDirective.prototype._component;
    /**
     * @type {?}
     * @private
     */
    AOContainerDirective.prototype._content;
    /**
     * @type {?}
     * @private
     */
    AOContainerDirective.prototype._resolver;
}
